### PR TITLE
Developing with SonarQube

### PR DESCRIPTION
**Pull request by EricKarpovits**
This PR introduces SonarQube, an invaluable addition to virtually all CSC301 projects. It offers an introduction to SonarQube, resources, compelling reasons for its adoption, a review of its limitations, and illustrative example, and getting started. I've aimed for simplicity without compromising depth to ensure it's an informative yet concise read. Please let me know if something is unclear and what should be changed according to the rubric they provided.

### CODE REVIEW

File: Topics/Development_Process/SonarQube.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+#### [Example Workflow](#example-workflow-1)
+#### [Getting Started](#getting-started-1)
+#### [Additional Resources](#additional-resources-1)
+
+## Introduction
+
+<img src="https://miro.medium.com/v2/resize:fit:666/1*rn-sO9oWLn9lYO7jkVO6og.png" width="300" height="100" align="right">
+
+[SonarQube](https://www.sonarsource.com/products/sonarqube/) is an open-source platform designed to continuously inspect code quality while easily integrating into the existing development workflow. It provides comprehensive code analysis and reporting, aiding developers and teams in maintaining code health, identifying bugs, vulnerabilities, and enhancing overall software quality. The solution performs checks at every stage of the development process. 
+
```
Comment (anoverga17) [inline Topics/Development_Process/SonarQube.md:16]: test

**Comment by anoverga17** [comment]
Excellent work! Some general comments and thoughts I had while reading this:

1. I am unsure how SonarQube benefits me as a developer if I am using a fully featured IDE (like IntelliJ). When first reading the advantages and disadvantages I thought to myself "doesn't intellij or a really customized VSCode do these things already?". I think code quality analysis and security vulnerability detection are especially things an IDE should have (though maybe not enabled by default). Feel free to correct me if I am wrong though as I don't explore everything my IDEs can do. This seems like its pretty resource intensive so knowing just how worth it it is to go through the hastle of setting it up and running it would be nice to know.

2. I think you need to make it a bit more clear on what SonarLint is and how it relates to SonarQube. When first reading this, I thought that SonarQube was what was hooked into your IDE when in actuality it was SonarLint. I would also like to know why would you install SonarLint if SonarQube does the same analysis later?

3. You mention maintenance as being an issue. As this tool runs on your device as a local server I am curious why that should be the case? Shouldn't updating it work like any other application? Should this tool ideally be deployed somewhere so your team does not have to individually run their own SonarQube server? (if so, you should probably mention that).

**Review by anoverga17 — COMMENTED** [review]
Also, some minor (more explicit) change suggestions:

1. There are uses of the term "we" throughout the article, which makes me think this is either an ad or parts of the article was copy and pasted from another source. You should probably replace these occurences with "SonarQube"

2. An image showing the code quality analysis would be nice. This seems to be SonarQube's competitive advantage so I would like to see how helpful the code analysis it suggests is.

Again, though, good writeup!

**Comment by EricKarpovits** [comment]
> Excellent work! Some general comments and thoughts I had while reading this:
> 
> 1. I am unsure how SonarQube benefits me as a developer if I am using a fully featured IDE (like IntelliJ). When first reading the advantages and disadvantages I thought to myself "doesn't intellij or a really customized VSCode do these things already?". I think code quality analysis and security vulnerability detection are especially things an IDE should have (though maybe not enabled by default). Feel free to correct me if I am wrong though as I don't explore everything my IDEs can do. This seems like its pretty resource intensive so knowing just how worth it it is to go through the hastle of setting it up and running it would be nice to know.
> 2. I think you need to make it a bit more clear on what SonarLint is and how it relates to SonarQube. When first reading this, I thought that SonarQube was what was hooked into your IDE when in actuality it was SonarLint. I would also like to know why would you install SonarLint if SonarQube does the same analysis later?
> 3. You mention maintenance as being an issue. As this tool runs on your device as a local server I am curious why that should be the case? Shouldn't updating it work like any other application? Should this tool ideally be deployed somewhere so your team does not have to individually run their own SonarQube server? (if so, you should probably mention that).

Hi Gabriel,

Thanks for your awesome comments, will definitely incorporate them in my next changes. 

Answering the questions:
1. The thing is IDEs tend to use open source and often times outdated plugins that just nearly aren't as good as what Sonar has to offer. 
IDEs tend to use linters and they don't generally look at the big picture and generally look at the immediate/nearby lines of code whereas SonarQube provides a comprehensive analysis of code quality, security vulnerabilities, and bugs across an entire project where things can arise through multiple files and coupling. Furthermore, SonarQube, offers centralized reporting and tracking of code quality metrics, making it easier for teams to collaborate and monitor improvements. Moreover, SonarQube offers diverse language support for the whole project; something like Intellij will only work for the java code and something else would have to be configed for other languages. Lastly, something really important is SonarQube allows for allows customization of rules and policies to match specific coding standards and best practices of teams and companies something once an IDE doesn't generally offer. Overall SonarQube is highly specific and meant for this that's why an IDE based approach would just not be suffice. Is this something I should clarify in my guide?
2. Okay I'll try to clarify it better. I might have a visual that can better describe it but essentially SonarLint is like the IDE plugin that looks at line by line very local and alerts you to clean up code as you code but SonarQube analyzes the big picture. 
3. Generally yes you would use SonarCloud, but its much more expensive and not something practical for csc301. I can put a blurb about it though.

**Comment by EricKarpovits** [comment]
> Also, some minor (more explicit) change suggestions:
> 
> 1. There are uses of the term "we" throughout the article, which makes me think this is either an ad or parts of the article was copy and pasted from another source. You should probably replace these occurences with "SonarQube"
> 2. An image showing the code quality analysis would be nice. This seems to be SonarQube's competitive advantage so I would like to see how helpful the code analysis it suggests is.
> 
> Again, though, good writeup!

Hi again, 
1. I looked through it and I use 'we' once in the following: 'We attain the following objectives:'. I might be wrong I thought it fit there as its an example workflow and its as if I'm going through it and afterwards get the following objectives. If I'm still wrong let me know and I'll definitely change it.
2. Sure will definitely add another image.  

Thanks again for your great comments, will definitely have them in mind as I finalize my guide.

**Comment by anoverga17** [comment]
>        The thing is IDEs tend to use open source and often times outdated plugins that just nearly aren't as good as what Sonar has to offer.
IDEs tend to use linters and they don't generally look at the big picture and generally look at the immediate/nearby lines of code whereas SonarQube provides a comprehensive analysis of code quality, security vulnerabilities, and bugs across an entire project where things can arise through multiple files and coupling. Furthermore, SonarQube, offers centralized reporting and tracking of code quality metrics, making it easier for teams to collaborate and monitor improvements. Moreover, SonarQube offers diverse language support for the whole project; something like Intellij will only work for the java code and something else would have to be configed for other languages. Lastly, something really important is SonarQube allows for allows customization of rules and policies to match specific coding standards and best practices of teams and companies something once an IDE doesn't generally offer. Overall SonarQube is highly specialized and is meant for this that's why an IDE based approach would just not be suffice. Is this something I should clarify in my guide?

Yeah, I think so. The guide was not very clear why I should use SonarQube over a default IDE or some other tool.

>  The thing is IDEs tend to use open source and often times outdated plugins that just nearly aren't as good as what Sonar has to offer.
IDEs tend to use linters and they don't generally look at the big picture and generally look at the immediate/nearby lines of code whereas SonarQube provides a comprehensive analysis of code quality, security vulnerabilities, and bugs across an entire project where things can arise through multiple files and coupling.

This first part clarified things immensely and might even be worth putting in the introduction

**Comment by anoverga17** [comment]
> I looked through it and I use 'we' once in the following: 'We attain the following objectives:'. I might be wrong I thought it fit there as its an example workflow and its as if I'm going through it and afterwards get the following objectives. If I'm still wrong let me know and I'll definitely change it.

Oh I see. I think the confusion is that for the example workflow, you mainly speak in a more second person tone. As an example 

> ...utilizing the Quality Gate profile **you've** set up to meet your acceptance criteria, SonarQube evaluates **your** PR

All of a sudden switching to we for this single part seems out of place. I think a better phrasing would be something along the lines of 'you attain the following objectives' or even better 'As you can see, by using SonarCloud, you gain the following benefits:'

**Pull request merged**